pkg/controller/restic_key_change_controller.go

```diff
@@ -0,0 +1,220 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+	informers "github.com/heptio/ark/pkg/generated/informers/externalversions/ark/v1"
+	listers "github.com/heptio/ark/pkg/generated/listers/ark/v1"
+	"github.com/heptio/ark/pkg/restic"
+)
+
+type resticKeyChangeController struct {
+	*genericController
+
+	namespace              string
+	secretLister           corev1listers.SecretLister
+	secretClient           corev1client.SecretsGetter
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter
+	resticRepositoryLister listers.ResticRepositoryLister
+	resticMgr              restic.RepositoryManager
+}
+
+// NewResticKeyChangeController creates a new restic key-change controller.
+func NewResticKeyChangeController(
+	logger logrus.FieldLogger,
+	namespace string,
+	secretInformer cache.SharedIndexInformer,
+	secretClient corev1client.SecretsGetter,
+	resticRepositoryInformer informers.ResticRepositoryInformer,
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter,
+	resticMgr restic.RepositoryManager,
+) Interface {
+	c := &resticKeyChangeController{
+		genericController:      newGenericController("restic-key-change", logger),
+		namespace:              namespace,
+		secretClient:           secretClient,
+		secretLister:           corev1listers.NewSecretLister(secretInformer.GetIndexer()),
+		resticRepositoryClient: resticRepositoryClient,
+		resticRepositoryLister: resticRepositoryInformer.Lister(),
+		resticMgr:              resticMgr,
+	}
+
+	c.syncHandler = c.processQueueItem
+	c.cacheSyncWaiters = append(c.cacheSyncWaiters, secretInformer.HasSynced, resticRepositoryInformer.Informer().HasSynced)
+
+	secretInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueIfKeyChange,
+			UpdateFunc: func(_, obj interface{}) {
+				c.enqueueIfKeyChange(obj)
+			},
+		},
+	)
+
+	return c
+}
+
+func (c *resticKeyChangeController) enqueueIfKeyChange(obj interface{}) {
+	secret := obj.(*corev1.Secret)
+
+	if _, ok := secret.Data[restic.ChangeCredentialsKey]; ok {
+		c.enqueue(obj)
+	}
+}
+
+func (c *resticKeyChangeController) processQueueItem(key string) error {
+	log := c.logger.WithField("key", key)
```

Comment:
To avoid confusion (I read the reassignment of `log` below and thought it was adding the `namespace` and `name` fields on top of `key`), how about if we remove this line and do `c.logger.WithField("key", key).Debug("Running processQueueItem")` here, and then below we can create `log` with the `namespace` and `name` fields?